modules/3d/include/opencv2/3d/detail/submap.hpp

```diff
@@ -0,0 +1,544 @@
+{
+    if (!points.needed() && !normals.needed())
+    {
+        MatType pts, nrm;
+        frame->getPyramidAt(pts, OdometryFrame::PYR_CLOUD, 0);
+        frame->getPyramidAt(nrm, OdometryFrame::PYR_NORM, 0);
+        volume->raycast(_cameraPose.matrix, intrinsics, frameSize, pts, nrm);
+        frame->setPyramidAt(pts, OdometryFrame::PYR_CLOUD, 0);
+        frame->setPyramidAt(nrm, OdometryFrame::PYR_NORM,  0);
+    }
+    else
+    {
+        volume->raycast(_cameraPose.matrix, intrinsics, frameSize, points, normals);
+    }
+}
+
+
+/**
+ * @brief: Manages all the created submaps for a particular scene
+ */
+template<typename MatType>
+class SubmapManager
+{
+   public:
+    enum class Type
+    {
+        NEW            = 0,
+        CURRENT        = 1,
+        RELOCALISATION = 2,
+        LOOP_CLOSURE   = 3,
+        LOST           = 4
+    };
+
+    struct ActiveSubmapData
+    {
+        Type type;
+        std::vector<Affine3f> constraints;
+        int trackingAttempts;
+    };
+
+    typedef Submap<MatType> SubmapT;
+    typedef std::map<int, Ptr<SubmapT>> IdToSubmapPtr;
+    typedef std::unordered_map<int, ActiveSubmapData> IdToActiveSubmaps;
+
+    SubmapManager(const VolumeParams& _volumeParams) : volumeParams(_volumeParams) {}
+    virtual ~SubmapManager() = default;
+
+    void reset() { submapList.clear(); };
+
+    bool shouldCreateSubmap(int frameId);
+    bool shouldChangeCurrSubmap(int _frameId, int toSubmapId);
+
+    //! Adds a new submap/volume into the current list of managed/Active submaps
+    int createNewSubmap(bool isCurrentActiveMap, const int currFrameId = 0, const Affine3f& pose = cv::Affine3f::Identity());
+
+    void removeSubmap(int _id);
+    size_t numOfSubmaps(void) const { return submapList.size(); };
+    size_t numOfActiveSubmaps(void) const { return activeSubmaps.size(); };
+
+    Ptr<SubmapT> getSubmap(int _id) const;
+    Ptr<SubmapT> getCurrentSubmap(void) const;
+
+    int estimateConstraint(int fromSubmapId, int toSubmapId, int& inliers, Affine3f& inlierPose);
+    bool updateMap(int _frameId, Ptr<OdometryFrame> _frame);
+
+    Ptr<detail::PoseGraph> MapToPoseGraph();
+    void PoseGraphToMap(const Ptr<detail::PoseGraph>& updatedPoseGraph);
+
+    VolumeParams volumeParams;
+
+    std::vector<Ptr<SubmapT>> submapList;
+    IdToActiveSubmaps activeSubmaps;
+
+    Ptr<detail::PoseGraph> poseGraph;
+};
+
+template<typename MatType>
+int SubmapManager<MatType>::createNewSubmap(bool isCurrentMap, int currFrameId, const Affine3f& pose)
+{
+    int newId = int(submapList.size());
+
+    Ptr<SubmapT> newSubmap = cv::makePtr<SubmapT>(newId, volumeParams, pose, currFrameId);
+    submapList.push_back(newSubmap);
+
+    ActiveSubmapData newSubmapData;
+    newSubmapData.trackingAttempts = 0;
+    newSubmapData.type             = isCurrentMap ? Type::CURRENT : Type::NEW;
+    activeSubmaps[newId]           = newSubmapData;
+
+    std::cout << "Created new submap\n";
```

Comment:
remove printfs / replace to logging functions.

Reply:
Done